File: 设计模式/程序/1-FactoryPattern/FactoryPattern(c#)/FactoryPattern.cs
/* 工厂模式C#实现 */
using System;

namespace ShapeFactory
{
    /* 接口 */
    interface Shape
    {   
        void draw();
    }

    /* 正方形 */
    class Rectangle: Shape
    {
        // 继承接口的函数不用加 override, 只有 vritual 与 override 搭配
        public /* override */ void draw()
        {
            Console.WriteLine("draw a Rectangle");
        }
    }

    /* 圆 */
    class Circle: Shape
    {
        public void draw()
        {
            Console.WriteLine("draw a Circle");
        }
    }

    /* 正方形 */
    class Square: Shape
    {
        public void draw()
        {
            Console.WriteLine("draw a Square");
        }
    }

    class ShapeFactory
    {
        /* 工厂函数 */
        public Shape GetShape(string shapeType)
        {
            if (shapeType == "Rectangle")
                return new Rectangle();
            else if (shapeType == "Circle")
                return new Circle();
            else if (shapeType == "Square")
                return new Square();
            return null;
        }

        static void Main(string[] args)
        {
            ShapeFactory shapeFactory = new ShapeFactory();
            Shape shape1 = shapeFactory.GetShape("Rectangle");
            Shape shape2 = shapeFactory.GetShape("Square");

            shape1.draw();
            shape2.draw();
        }
    }
}

File: 设计模式/程序/1-FactoryPattern/FactoryPattern(c++)/FactoryPattern.cpp
#include <iostream>
#include <string>

/* C++实现工厂模式 */
namespace ShapeFactory
{
    /* 接口类 */
    class Shape
    {
    public:
        virtual void draw() = 0;
        virtual Shape(){}
    };

    /* 正方形类 */
    class Rectangle: public Shape
    {
    public:
        virtual void draw() override
        {
            std::cout << "draw a Rectangle" << std::endl;
        }
    };

    /* 圆形 */
    class Circle: public Shape
    {
    public:
        virtual void draw() override
        {
            std::cout << "draw a Circle" << std::endl;
        }
    };

    /* 正方形 */
    class Square: public Shape
    {
    public:
        virtual void draw() override
        {
            std::cout << "draw a Square" << std::endl;
        }
    };

    /* 工厂函数 */
    Shape* GetShape(const std::string& shapeType)
    {
        if (shapeType == "Rectangle")
            return new Rectangle();
        else if (shapeType == "Circle")
            return new Circle();
        else if (shapeType == "Square")
            return new Square();
        return nullptr;
    }
}


int main()
{
    /* 使用工厂函数 */
    ShapeFactory::Shape* shape1 = ShapeFactory::GetShape("Rectangle");
    ShapeFactory::Shape* shape2 = ShapeFactory::GetShape("Circle");

    shape1->draw();
    shape2->draw();

    delete shape1;
    delete shape2;

    return 0;
}
